### PR TITLE
Ingestor failure/warning recording

### PR DESCRIPTION
**Pull request by ironfroggy**

### CODE REVIEW

File: voter/management/commands/voter_process_snapshot.py [hunk 1]
```diff
@@ -126,94 +151,135 @@ def reset_file(file_tracker):
...
+    # If we see a repeat, flushed queued data before continuing
+    # This prevents the same voter from appearing twice in a single bulkd insert
+    ncid = row.get('ncid')
+    if ncid in processed_ncids:
+        flush()
+    voter_instance = find_existing_instance(ncid)
+
+    # Skip rows that have no NCID in them :-(
+    # TODO: Log these so we can come back and figure out what to do with them
+    if ncid is None:
```
Comment (vkurup) [inline voter/management/commands/voter_process_snapshot.py]: Would it be appropriate to change this to `if not ncid`? We'd want to raise the error for empty strings as well.

**Comment by vkurup** [comment]
My voter process just errored out with this:
```
Processing NCVoter file...
Tracking changes for file /media/ntfsdrive/downloads/ncvoter/2018-03-16T16:11:03:1521216663/VR_20051125.txt
3399061it [2:23:20, 395.20it/s]
Traceback (most recent call last):█████████████████████████████▌                                                                                   | 3399061/8238290.943843842 [2:23:20<3:24:05, 395.20it/s]
  File "/home/vkurup/dev/voters-ingestor/voter/management/commands/voter_process_snapshot.py", line 267, in track_changes
    ncid, voter_instance = skip_or_voter(row)
  File "/home/vkurup/dev/voters-ingestor/voter/management/commands/voter_process_snapshot.py", line 197, in skip_or_voter
    raise ValueError("No NCID found in data")
ValueError: No NCID found in data

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
    return self.cursor.execute(sql, params)
psycopg2.IntegrityError: duplicate key value violates unique constraint "voter_badline_filename_line_no_d6a51488_uniq"
DETAIL:  Key (filename, line_no)=(/media/ntfsdrive/downloads/ncvoter/2018-03-16T16:11:03:1521216663/VR_20051125.txt, 3399084) already exists.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/vkurup/dev/voters-ingestor/voter/management/commands/voter_process_snapshot.py", line 321, in process_files
    added, modified, ignored, skipped = track_changes(file_tracker, output)
  File "/home/vkurup/dev/voters-ingestor/voter/management/commands/voter_process_snapshot.py", line 271, in track_changes
    BadLine.error(file_tracker.filename, total_lines, line, e.args[0])
  File "/home/vkurup/dev/voters-ingestor/voter/models.py", line 55, in error
    is_warning = False,
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/query.py", line 417, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/base.py", line 729, in save
    force_update=force_update, update_fields=update_fields)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/base.py", line 759, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/base.py", line 842, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/base.py", line 880, in _do_insert
    using=using, raw=raw)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/query.py", line 1125, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1281, in execute_sql
    cursor.execute(sql, params)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/backends/utils.py", line 100, in execute
    return super().execute(sql, params)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
    return self.cursor.execute(sql, params)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.IntegrityError: duplicate key value violates unique constraint "voter_badline_filename_line_no_d6a51488_uniq"
DETAIL:  Key (filename, line_no)=(/media/ntfsdrive/downloads/ncvoter/2018-03-16T16:11:03:1521216663/VR_20051125.txt, 3399084) already exists.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
    utility.execute()
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/core/management/__init__.py", line 365, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/vkurup/.virtualenvs/ncvoter/lib/python3.6/site-packages/django/core/management/base.py", line 335, in execute
    output = self.handle(*args, **options)
  File "/home/vkurup/dev/voters-ingestor/voter/management/commands/voter_process_snapshot.py", line 351, in handle
    process_files(output=True)
  File "/home/vkurup/dev/voters-ingestor/voter/management/commands/voter_process_snapshot.py", line 324, in process_files
    raise Exception('Error processing file {}'.format(file_tracker.filename))
Exception: Error processing file /media/ntfsdrive/downloads/ncvoter/2018-03-16T16:11:03:1521216663/VR_20051125.txt
(ncvoter) vkurup@caktus020:~/dev/voters-ingestor(ingestor-logging) $ 
```

Somehow it's violating a unique constraint?

**Comment by ironfroggy** [comment]
@vkurup did you happen to try to import a file more than once? It looks like it failed because it hit the same error from the same file line twice. Should have thought of that. Will corrrect.

**Comment by vkurup** [comment]
Hmm, this was my first time trying to import... but I'll try again with this commit

File: voter/management/commands/voter_process_snapshot.py [hunk 2]
```diff
@@ -44,20 +52,42 @@ def find_md5(row_data, exclude=[]):
...
         f.seek(0)
         if f.read(2) == b'\xff\xfe':
             encoding = 'utf16'
         else:
             encoding = 'latin1'
+        return encoding
+
+
+def get_file_lines(filename, output):
+    tqdm = tqdm_or_quiet(output) # noqa, F811
```
[the remark this answers is not in the document]
Comment (ironfroggy) [reply voter/management/commands/voter_process_snapshot.py]: Hmm... i _was_ getting an error about shadowing a global, but i don't seem to get it now. I'm not sure why, but I've removed it now.

File: voter/models.py [hunk 3]
```diff
@@ -109,34 +142,35 @@ class Meta:
...
+            parsed_row['birth_year'] = int(raw_birth_year)
 
         snapshot_dt = row.get('snapshot_dt')
         if snapshot_dt:
             snapshot_dt = snapshot_dt[:10]
             snapshot_dt = datetime.strptime(snapshot_dt, '%Y-%m-%d').replace(tzinfo=pytz.timezone('US/Eastern'))
-            row['snapshot_dt'] = snapshot_dt
+            parsed_row['snapshot_dt'] = snapshot_dt
 
         return row
```
[the remark this answers is not in the document]
Comment (ironfroggy) [reply voter/models.py]: yes,

mutating the row in place was causing some trouble writing a clean test in some case

**Pull request merged**